mlir/include/mlir/IR/DialectRegistry.h

```diff
@@ -255,7 +256,9 @@ class DialectRegistry {
 
 private:
   MapTy registry;
-  std::vector<std::unique_ptr<DialectExtensionBase>> extensions;
+  using KeyExtensionPair =
+      std::pair<TypeID, std::unique_ptr<DialectExtensionBase>>;
+  llvm::SmallVector<KeyExtensionPair> extensions;
```

Comment:
If the overall theme of this change looks good, I'd like to also convert this into a `DenseMap` so we always de-duplicate by key (instead of doing that implicitly in code paths that call `isSubset`).

It'll make `applyExtensions` slightly less efficient since we'll have to track changes in case additional extensions are registered when applying the existing extensions set, but I think the code clarity would probably be worth it.

Reply:
You can't use a DenseMap here, that will make iteration order non-deterministic, this could be a MapVector though (which still gives you random access via the begin() iterator).

Reply:
Ah, I didn't realize iteration order was important. Switched to MapVector!